DESIGN.md

```diff
@@ -0,0 +1,111 @@
+# Non-Blocking Alter Table Support (SqlTable)
+
+## Overview
+The overall goal of this project is to implement lazy, non-blocking schema changes. The current storage layer does not support any schema changes. The first goal of this project is to support schema change operations including add column, drop column, and default value changes. However, the overall goal would be to carry this out in a non-blocking fashion, for which we decided to go with a lazy evaluation approach. This means that any schema change would not migrate existing tuples to the new schema until information that only exists in the new schema is modified.
+
+## Scope
+Almost all of the work will be localized to the SqlTable object as that is the access point from the execution layer to the storage layer. Also our design will not affect the underlying structures of the tuples or the DataTables and after this change they should still be able to provide their functionality without having to go through the SqlTable. Within the SqlTable we will be changing the following modules. Below we will refer to expected version as the version of the tuple the user expects to see and actual version as the version the tuple is currently in. These two versions will differ when a user has updated schema, but due to the lazy evaluation the tuple hasn’t been transformed into the latest version.
+
+### SqlTable::Insert
+- Insert will now take in a schema version number that indicates the version of the tuple being inserted.
+- In this case the actual and expected versions will be the same as an insert will always put a tuple in the version that is passed in.
+
+### SqlTable::Update
+- Update will be required to take in a schema version number that indicates the expected schema version number of the tuple being updated
+- In this case the actual version number will differ from the expected version in cases where the tuple being updated has not been shifted to the expected version.
+- If the update modifies columns that are not in the actual version then the tuple will be shifted to the expected version before applying the update.
+
+### SqlTable::Scan
+- Scan will be required to take in a schema version number that indicates the schema version number that the current transaction sees.
+
+### SqlTable::Delete
+- Delete will be required to take in the schema version number that indicates the expected schema version number of the tuple being deleted
+
```

Comment:
You also modified `InitializerForProjectedColumns`, `InitializerForProjectedRows` .etc. You maybe also need to document them here.